composer/trainer/trainer.py

```diff
@@ -1091,8 +1081,94 @@ def _train_loop(self) -> None:
+                               This means the GPU does not have enough memory to process even 1"""))
+        else:
+            self.state.grad_accum = min(2 * self.state.grad_accum, self.state.batch_num_samples)
+            self.logger.data_batch({'trainer/grad_accum': self.state.grad_accum})
+
+    def _compute_metrics(self, train_metrics: Union[MetricCollection, None]):
+        """Compute training metrics.
+
+        Args:
+            train_metrics (Union[MetricCollection, None]): Existing train metrics
+        """
+        # Compute metrics on the training set
+        if self._compute_training_metrics:
+            assert train_metrics is not None
+            self.state.model.eval()
+            with torch.no_grad():
+                for eval_microbatch in self._train_data_spec.split_batch(self.state.batch, self.state.grad_accum):
+                    # TODO: Detect if self.run_event(Event.AFTER_DATALOADER) changes the training
+                    # data and if so print a warning that metrics may return unexpected results
+                    outputs, targets = self._original_model.validate(eval_microbatch)
+                    train_metrics.update(outputs, targets)
+
+    def _train_batch(self, use_grad_scaling: bool):
+        """Compute loss by training on a full batch of data. Adaptively change microbatch size if enabled to maximize
+        GPU usage.
+
+        Args:
+            use_grad_scaling (bool): Enables gradient scaling
+        """
+        # Retry until we successfully complete training and return loss
+        while True:
+            total_loss = None
+            should_handle_cuda_oom = False
+            caught_error = None
+            try:
+                assert self.state.scaler is not None
+                microbatches = self._train_data_spec.split_batch(self.state.batch, self.state.grad_accum)
+                if self.deepspeed_enabled:
+                    total_loss = self._train_microbatches(microbatches)
+                elif self._use_closures():
+                    for optimizer in self.state.optimizers:
+                        if use_grad_scaling:
+                            total_loss = self.state.scaler.step(
+                                optimizer, closure=lambda **kwargs: self._train_microbatches(microbatches, **kwargs))
+                        else:
+                            total_loss = optimizer.step(
+                                closure=lambda **kwargs: self._train_microbatches(microbatches, **kwargs).item())
+                else:
+                    total_loss = self._train_microbatches(microbatches)
+                    for optimizer in self.state.optimizers:
+                        if use_grad_scaling:
+                            self.state.scaler.step(optimizer)
+                        else:
+                            optimizer.step()
+            except RuntimeError as e:
```

Comment:
I'm not super familiar with this part, so correct me if I'm wrong. My understanding is it would hit a timeout (set with `init_process_group`) and surface an error. When this error is surfaced, it's caught across all ranks. We then hit the `except` here and recognize it was really driven by an OOM, which we appropriately handle. So, the situation is still handled

Reply:
As a side, I'm fine adding logging, but note that it will fire every time we hit an auto `grad_accum` update. Personally think it's better to not log that often with potential issues since it drowns out stuff.

Reply:
Ah yes -- I missed the logic with the timeout. If rank 1 times out, then yup -- agreed it would hit here, and handle it appropriately.

Maybe could you add this as a `log.info` or `log.debug`? I think by default python only prints `log.warn`, but if a user runs into an issue, then we'll ask them to include the log output when running on debug mode, so it would be helpful to include this in the stack trace. I don't think there will be too many grad accum updates -- just a few at the beginning of training?

(In general, we need to add more logging statements to the training loop! Don't have enough IMO).

Reply:
Got it -- will add more logging :)